Simplify operations

jech opened the pull request:
This simplifies the operations queue.  While the new code should be slightly faster and more robust, the main benefit is to reduce the amount of noise in the blocking profile.

Sean-Der commented on operations.go:
```diff
@@ -9,60 +9,73 @@ type operation func()
 }
 
 func newOperations() *operations {
 	closed := make(chan struct{})
```
Can you delete this and the following line, thanks!

Sean-Der commented on operations.go:
```diff
@@ -9,60 +9,73 @@ type operation func()
+
 	o.mu.Lock()
-	defer o.mu.Unlock()
+	busy := o.busy
```
What do you think of something like `queueInProgress`? Major nit, but for some reason I kept having to re-read because `busy` is on the struct and on stack.

Sean-Der commented on operations.go:
```diff
@@ -9,60 +9,73 @@ type operation func()
-	o.mu.Lock()
-	defer o.mu.Unlock()
-	return o.done
+// Done blocks until all currently enqueued operations are finished executing.
```
Mind returning a `context.Context`, then the caller can give up waiting if they want. In the future might be useful for timeouts/measuring stuff.

Sean-Der commented:
This fantastic! Thank you so much @jech, if you disagree I am happy to just push things forward. 

I am just paranoid about adding a bug to the operations queue, seem like the best place for a deadlock to pop up

jech commented:
> Can you delete this and the following line, thanks!

Done, sorry.

> What do you think of something like queueInProgress? Major nit, but for
> some reason I kept having to re-read because busy is on the struct and
> on stack.

Renamed to `running`.  Better?

> Mind returning a context.Context

I disagree, for at least two reasons.  First, any more complex
synchronisation can be easily built by the user; I've added an example in
operations_test.go.  Second, many people dislike Context, and I don't
think we should be imposing its use.

> I am just paranoid about adding a bug to the operations queue, seem like
> the best place for a deadlock to pop up

I agree, we should get rid of it altogether.

-- Juliusz

Sean-Der commented:
LGTM! Merging, thank you so much @jech

Sean-Der commented:
Merged with 9bbffdd5aab2f271629aef6b912190d0de358d6e thank you!
